hugo/content/posts/nix-bazel-crosspiling.md

```diff
@@ -0,0 +1,129 @@
+
+- Having to switch developer machines and synchronize code across them.
+- Having to rig up together some form of individual remote compilation and teach everyone how to use it (for example, having Windows sync the repo to Mac, run the build on Mac, and report back the results).
+- Relying on continuous integration to build non-host platforms and not having coverage during normal developer iteration.
+
+Conversely, I see some major upsides to crosspiling:
+
+- Running "build all" can actually build **all**.
+- Developer environments can be standardized to a single development environment (like a Linux [DevServer]({{< ref "/posts/nixos-devserver.md" >}})).
+- Automation infrastructure (like continuous integration) can also be standardized to the same single development environment.
+
+With these motivations in mind, the question becomes, how do I actually set this up? From the reading and searching I've done, I had a hard time finding a concise and clear example of crosspiling "from scratch". This was particularly evident with my preferred choice of package management and build system, Nix and Bazel respectively. Hopefully this post can help serve as such an example.
+
+For background information and motivation on why Nix + Bazel, I recommend reading <https://www.tweag.io/blog/2018-03-15-bazel-nix/>.
+
+# Nix
+
+Nix is the package manager providing all the tools and SDKs needed to crosspile code. To pull this off, my [Nix Flake](https://github.com/ConsumingChaos/consumingchaos.github.io/tree/master/examples/nix-bazel-crosspiling/flake.nix) exposes two things, a `devShell` (making tools available in the terminal and VSCode) and a set `bazel` that contains derivations that get exposed to Bazel via [rules_nixpkgs](https://github.com/tweag/rules_nixpkgs). Keeping everything in the same Nix Flake lets me ensure there is only one set of tools and SDKs and that they're shared between the developer environment and Bazel.
+
+The main derivation is `bazel.config` which generates `config.bzl`, a Bazel file containing all of the Nix store paths I need exposed to Bazel. I opted to use absolute paths (where possible) for Nix provided packages because they are already content addressed, so any change to them will result in a new path, which in turn will invalidate Bazel's cache. Unfortunately this won't work with Bazel remote execution, but `rules_nixpkgs` doesn't support remote execution presently anyways. On the flip side, this skips the need for Bazel to make a giant symlink forest for each package, and in the case of some of the SDKs, works around the issue where they contain recursive symlinks, which Bazel chokes on.
+
+## SDKs
+
+Below are my rough instructions for acquiring the SDKs to each platform targeted in this post. The Apple and Microsoft ones are not publicly hostable, so it's left up to the reader to solve making them privately available for their project.
```

Comment:
I would clean up this sentence:

"The Apple and Microsoft SDKs cannot be hosted publicly, so it's left up to the reader to make them privately available for their project."